### ISSUE TITLE
spack install --all-variants (for testing)

### ISSUE THREAD

**Issue by markcmiller86** [LLNL/spack#4530]
I have a package with many variants. How do I test all the logic I've added for each variant and (appropriate) combinations thereof and/or error detection of inappropriate combinations? Only option at present is manually invoking various `spack install` operations.

But, since spack knows all the variants for a package, a useful testing ability before submitting a pull request for a package would be a `spack install --all-variants`. I would probably do it also with `--fake` also but no need to assume that either.

**Comment by adamjstewart** [comment]
Better combinatorial testing would also be useful for @tgamblin's planned package build testing once CDash is up and running.

**Comment by davydden** [comment]
For 2-3 variants that's ok, but for 10-20 u would have to test manually the most used combos as the total number of combination is insane. IMO this feature would not be of practical significance

**Comment by markcmiller86** [comment]
Manually, really? Thats the answer when I am sitting in front of a machine *designed* to automate things? That isn't very satisfying.

Maybe have option to codify the "most used" variants in the package.py file and have spack run through them with `--all-variants`. Otherwise, I am beginning to doubt the value of my current practice of trying, to extent time is available, to export to spack many of the configuration options I find available in packages I am working with.

**Comment by davydden** [comment]
Yes, manually. 2^10=1024 combinations, if it's something like Trilinos which takes good 30 mins, u r looking at 15k hours. It's ok to test with all 'on' and polish the rest while using.

**Comment by markcmiller86** [comment]
Yeah, I can do the math. I don't like the manual part though. If what you propose is to do something with some subset of variants (which is what I understood you to mean when you said "...would have to test manually the most used combinations), then I propose that thing be automated, not manual and we include that list of most used combinations *in* the package.py file for the package and then have `spack --install --most-used-variants` or some such to *automate* that process. I mean, I like the idea of subsetting the variants tested. I don't like the idea of maintaing that list apart from spack, or manually processing it.

**Comment by markcmiller86** [comment]
BTW...do we have an easy way to "...test with all 'on'..." for a package even? I did this command...
```
[scratlantis:~/spack/spack] miller86% ls ~/spack/spack/var/spack/repos/builtin/packages/*/*.py | xargs -n 1 -I '{}' sh -c "grep variant {} | wc -l" | sort | uniq
       0
       1
       2
       3
       4
       5
       6
       7
       8
       9
      10
      11
      12
      13
      14
      16
      17
      21
      22
      23
      29
      43
```
For those packages with say more than 5 variants, achieving the state of all 'on' is a non-trivial command-line. Maybe a way to do that as well would be useful for spack to offer.

**Comment by davydden** [comment]
I also don't like the manual part.

**Comment by davydden** [comment]
Ps. Doing all combinations among variants would be good to have, I agree. But this s should also cover compilers and possibly dependencies.

**Comment by tgamblin** [comment]
With the new concretizer we will have a way to iterate through all solutions, which would cover this case, but we'd need to provide a syntax for what to explode combinatorially and what not to.  The test-suite file does some of that, but it needs some support for iterating over legal combinations

**Comment by markcmiller86** [comment]
> @davydden ...but this should also cover compilers and possible dependencies...
I guess. But, my main concern is to test the package code I have written. For example, I just discovered I had spelled `threadsafe` as `theadsafe` (missing the `r`) in MFEM pull request. As an aside, even a simple spell check comparison of literal strings used in `variant()`, `depends_on()` relative to all other literal strings in the package.py file would be helpful in catching those kinds of problems.
